### ISSUE TITLE
[MechanicalObject] Non-mapped MechanicalObjects influencing the solvers

### ISSUE THREAD

**Issue by marques-bruno** [sofa-framework/sofa#581]
Hi,
When adding a MechanicalObject in a subnode of one containing a solver, its positions are taken into account by the solver (addKToMatrix) even if it is not mapped to the MechanicalObject in the solver's node.

I am not sure that this makes sense, but if it does, a little warning message seems necessary to avoid spending time (hours in my case, as physics is not my strong suit...) searching for what goes wrong in the scene.
Looking fwd to your input on this topic!
______________________________________________________

Suggested labels:

**Comment by guparan** [comment]
Hey @marques-bruno, any update on this issue?

**Comment by marques-bruno** [comment]
Hi!
Here's a simple test scene showing the problem:
```<?xml version="1.0" ?>
<Node name="lroot" gravity="0 -9.81 0" dt="0.02">
    <Node name="Liver">
        <EulerImplicitSolver name="cg_odesolver"   rayleighStiffness="0.1" rayleighMass="0.1" />
        <SparseLDLSolver  saveMatrixToFile="true"/>
        <!-- <CGLinearSolver name="linear solver" iterations="1000" tolerance="1e-09" threshold="1e-09" /> -->
        <MeshGmshLoader name="meshLoader" filename="mesh/liver.msh" />
        <TetrahedronSetTopologyContainer name="topo" src="@meshLoader" />
        <MechanicalObject name="dofs" src="@meshLoader" />
        <TetrahedronSetGeometryAlgorithms template="Vec3d" name="GeomAlgo" />
        <DiagonalMass  name="computed using mass density" massDensity="1" />
        <TetrahedralCorotationalFEMForceField template="Vec3d" name="FEM" method="large" poissonRatio="0.3" youngModulus="3000" computeGlobalMatrix="0" />
        <FixedConstraint  name="FixedConstraint" indices="3 39 64" />
        <Node name="BUG" >
          <MechanicalObject name="MO" position="1 1 1 2 2 2 3 3 3" />
          <Sphere radius="1" />
        </Node>
    </Node>
</Node>
```

I have a very bad comprehension of these things, but apparently, Sofa concatenates the points in the mechanical object "/dofs" and "/BUG/MO", even though there is no mapping to link both objects.
If you replace the SparseLDLSolver by the CGLinearSolver, it will work, but that's because CG doesn't invert the matrix.
I am reporting @IPeterlik's explanations here, but Igor can undoubtedly explain better than me what's actually happening here.. :)

**Comment by IPeterlik** [comment]
Hello, in my opinion, this bug is related to the actual status of solvers in SOFA. Since CG is a very "robust"  solver (where "robust" means that it constructs solutions which are mathematically not always justifiable), it is used everywhere in examples. However, there are other solvers available in SOFA (even if not used often), such as SparseLDL, SparseLU. I think it might be highly "surprising" for someone that while CG calculates some solution (which IMHO is not mathematically justifiable, but let's put this aside), LDL or LU will simply crash on such as simple scene without saying anything clear about the reason of crashing. Or to put it differently -- the above scene is not well-formed (a non-mapped MO in a subnode...), but if using CG, everything looks fine (even IMHO it really should not).

**Comment by IPeterlik** [comment]
And one more note --  with this scene, if using CG, it is even impossible to learn what is happening during the calculations. If you use a direct solver capable of exporting the matrix, you will learn that in the above scene, the size of the system matrix is  3*184 X 3*184. Thus mathematically speaking, the two mechanical systems were _concatenated_ (181 points in master node + 3 nodes in slave node) which is IMHO an absolute nonsense. But again, you'll never learn that directly from SOFA (only if you examine the matrix dumped by LDL before it crashes).

**Comment by damienmarchal** [comment]
Hi all,
+1 to Igor and I add @olivier-goury to the discussion as he is very interested by the topic. 

1) Having SparseLDL crashing is a shame (or a joke) that was during since too long. The good news is that the issue was fixed last week in this PR: 
- https://github.com/sofa-framework/sofa/pull/655 (it shouldn't crash that much :))

2) About the lack of get messages... I discovered there is a  #define MULTIMATRIX_VERBOSE that can be changed and this is actually really helpfull to get a minimal amount of understanding of what is happening. https://github.com/sofa-framework/sofa/pull/660/commits/90f413eb2b79ba45fbc742e9777691203ca33886

Now having to change a #define in some obscure .h file is also a shame so as a first step we should replace this #define with a real variable that users can activate/deactivate at run time (without recompiling).  This is easy to do. 

3) At Defrost those aspects are under heavy investigation and we are probably willing to help anyone that comes with a road map to improve the situation. 

DM.
